Overhaul Word Select Pane

cofinley opened the pull request:
Move to QTextBrowser + HTML instead of a QLabel object for each 
word/token. This should decrease resource consumption and make
the text look cleaner. The QLabels with large margins made the text
look less readable.

Use cleaner methods of differentiating actual text from punctuation,
spaces, symbols, etc. The new methods are more correct, as I found
out when trying a longer French text. The apostrophes making the
elisions and hyphenated words are poorly parsed. All of these factors
made the word select look bad, m'kay.

Before:

![](https://i.imgur.com/YGAB9Ps.png)

After: 

![](https://i.imgur.com/naM5zPV.png)

The new look is a bit simpler, but the subset of CSS allowed in Qt stylesheets doesn't allow for more padding. On the plus side, I think it looks/reads more like the original text.

---

P.S. Remove the symbols/punctuation from the default ignored_words in config.
These won't be a problem from now on since they aren't considered in the
model of valid words that are clickable.

cofinley committed:
Overhaul word select pane

Move to QTextBrowser + HTML instead of a QLabel object for each word/token.
This will hopefully dramatically decrease resource consumption and make
the text look cleaner. The QLabels with large margins made the text
look less readable.

Use cleaner methods of differentiating actual text from punctuation,
spaces, symbols, etc.

Remove the symbols/punctuation from the default ignored_words in config.
These won't be a problem from now on since they aren't considered in the
model of valid words that are clickable.

cofinley commented:
## Measurements

Rough memory usage measurements using `ps` and the increase in resident set size (rss) memory (also rough).

### Small Example

Paragraph, 150 words, 0.8 KB of plaintext

- Old word select
  - Memory usage: 8160 KB
  - Load time: near-instant
- New word select
  - Memory usage: 7568 KB
  - Load time: near-instant
- Savings
  - Space: 500 KB (~6%)
  - Time: none

### Large Example

Chapter, 3500 words, 21 KB of plaintext

- Old word select
  - Memory usage: 26456 KB
  - Load time: 13 seconds, terrible lag on scroll
- New word select
  - Memory usage: 20356 KB
  - Load time: 6 seconds, **no lag on scroll**
- Savings
  - Space: 6 MB (~25%)
  - Time: 6 seconds (**50%**)

jeprecated commented on src/find_missing_words/gui/search.py:
```diff
@@ -128,17 +126,20 @@ def search(self):
             config = mw.addonManager.getConfig(__name__)
             ignored_words = config[ConfigProperties.IGNORED_WORDS.value]
-            known = len(found_note_ids) > 0 or token.lower() in [ignored_word.lower() for ignored_word in ignored_words]
+            is_word = utils.is_word(token)
```
Repeated effort, move variable to top of for loop so it is not done twice. Actually... :thinking: if this is false it'll never get into here anyways :100: :thinking:

cofinley replied:
Good catch, wrote this line before the other one and forgot about it.

cofinley replied:
I'll also add some comments!

jeprecated reviewed (commented):
Minimal comments looking through the code itself :+1:

ohare93 commented:
1) When I select a word my word window goes half blank :thinking: 

![image](https://user-images.githubusercontent.com/1471350/69572705-03814980-0fc5-11ea-9ff4-6dac1ea5ae21.png)

Goes away when I click again.

Double clicking seems to try and select the whole box, which does something similar but different!

![image](https://user-images.githubusercontent.com/1471350/69572807-2f9cca80-0fc5-11ea-91e5-4cf18f00cf33.png)



1) Why does the "Notes Created" tab only get filled with entries individual to the specific word you have chosen? :thinking: I made three new words, and that view only shoes one entry. Which is the card I made, and changes to show each of them upon clicking the card again. I assumed it would be a list of all the cards. Is it not?

cofinley commented:
> 1. Why does the "Notes Created" tab only get filled with entries individual to the specific word you have chosen? 🤔 I made three new words, and that view only shoes one entry. Which is the card I made, and changes to show each of them upon clicking the card again. I assumed it would be a list of all the cards. Is it not?

Hmm, not sure what you mean. Can you provide some steps to reproduce this?

Possible problem: did you change the destination deck when adding some of the notes? I noticed that if you change the deck when adding, it won't show up on the notes list once it's created b/c it's technically not found with the original deck filter (if you set one). For example: searching with "Default"as the filter deck, starting a new note from a preset, setting it to another deck, saving the new note, then seeing that the note does not appear in the note list when you click on the word again.

Interesting with the blank screen. I can try Windows, but my Ubuntu was working fine.

cofinley committed:
Fix issue #10, add comments to new word select code

ohare93 commented:
Sorry, should have replied sooner. Here's a video showing what I mean https://1drv.ms/v/s!An8TBTYs7ARHh_s7mNaZO7tp7wRVEA?e=H4ortJ it also shows some weird graphical issues I've been having :sweat: 

On an unrelated note: what program do you use to record things to gifs on Linux :tired_face:

cofinley committed:
Improve readability of word select pane

Highlighted words were looking too cramped, needed padding.
Regular padding not possible unless using <table> in Qt[0], use html
spacing entity to fake the padding within <a> tags.

Respect paragraph breaks (double-spacing/newline between words).

[0]: https://doc.qt.io/qt-5/richtext-html-subset.html#css-properties

cofinley commented:
> 
> 
> Sorry, should have replied sooner. Here's a video showing what I mean https://1drv.ms/v/s!An8TBTYs7ARHh_s7mNaZO7tp7wRVEA?e=H4ortJ it also shows some weird graphical issues I've been having 😓
> 
> On an unrelated note: what program do you use to record things to gifs on Linux 😫

I use [ShareX](https://getsharex.com/) on Windows and record my Ubuntu VM window :)

Thanks for the video. That is very odd behavior.... What version of Anki/PyQt are you using? In Anki>Help>About...

I have:
```
Version 2.1.8 (71e0c880)
Qt 5.12.2 PyQt 5.12.1
```

cofinley committed:
Use OS-independent pt measurement in word select CSS

ohare93 commented:
> I use ShareX on Windows and record my Ubuntu VM window, but I've heard good things about OBS, which is available on Linux.

Cool, thanks

Here's my version info:

```
Version 2.1.0beta36
Qt 5.9.5 PyQt 5.10.1
```

cofinley commented:
Tried on the latest beta build of 2.1.17 on Kubuntu and could not reproduce. Hard to fix without more diagnostic info. Could be that your Qt/PyQt is behind? Merging and closing for now in order to move forward with the overhaul.
